code_map.py

```diff
@@ -1,296 +1,346 @@
-def code_map_file():
 
-    plugin_dir = ''
+# -------------------------
 
-    if hasattr(sublime, 'cache_path'):
-        plugin_dir = sublime.cache_path()
-    else:
-        plugin_dir = 'cache'
-        plugin_dir = os.path.join(os.getcwd(), plugin_dir)
 
-    data_dir = path.join(plugin_dir, 'CodeMap', 'CodeMap.Data', str(sublime.active_window().id()))
-    if not path.exists(data_dir):
-        os.makedirs(data_dir)
-    return path.join(data_dir, 'Code - Map')
-# -----------------
-def set_layout_columns(count, coll_width=0.75):
+def code_map_file():
 
-    if count == 1:
-        sublime.active_window().run_command("set_layout", {"cells": [[0, 0, 1, 1]], "cols": [0.0, 1.0], "rows": [0.0, 1.0]})
+    dst = path.join(packages_dir, 'User', 'CodeMap')
+    return path.join(dst, 'Code - Map')
 
-    elif count == 2:
-         sublime.active_window().run_command("set_layout", {"cells": [[0, 0, 1, 1], [1, 0, 2, 1]], "cols": [0.0, coll_width, 1.0], "rows": [0.0, 1.0]})
+# -----------------
 
-    elif count == 3:
-         sublime.active_window().run_command("set_layout", {"cells": [[0, 0, 1, 1], [1, 0, 2, 1], [2, 0, 3, 1]], "cols": [0.0, 0.33, 0.66, 1.0], "rows": [0.0, 1.0]})
 
-    elif count == 4:
-         sublime.active_window().run_command("set_layout", {"cells": [[0, 0, 1, 1], [1, 0, 2, 1], [2, 0, 3, 1], [3, 0, 4, 1]], "cols": [0.0, 0.25, 0.5, 0.75, 1.0], "rows": [0.0, 1.0]})
-# -----------------
-def centre_line_of(view, region):
-    (first_row,c) = view.rowcol(region.begin())
-    (last_row,c) = view.rowcol(region.end())
-    return int(first_row + (last_row - first_row)/2)
-# -----------------
 def get_code_map_view():
-    for v in sublime.active_window().views():
+    for v in win().views():
         if v.file_name() == code_map_file():
             return v
+
 # -----------------
+
+
 def refresh_map_for(view):
     file = view.file_name()
-    if code_map_generator.can_map(file):
+    if file and code_map_generator.can_map(file):
         code_map_view = get_code_map_view()
         if code_map_view:
-            code_map_view.run_command('code_map_generator', {"source": file })
+            code_map_view.run_command('code_map_generator', {"source": file})
 
-# ===============================================================================
-class event_listener(sublime_plugin.EventListener):
-    map_closed_group = -1
-    pre_close_active = None
-    can_close = False
-    # -----------------
-    def on_load(self, view):
-        if view.file_name() != code_map_file():
-            refresh_map_for(view)
-    # -----------------
-    def on_pre_close(self, view):
-        if view.file_name() == code_map_file():
-            event_listener.map_closed_group, x = sublime.active_window().get_view_index(view)
-            if len(sublime.active_window().views_in_group(event_listener.map_closed_group)) == 1:
-                event_listener.can_close = True
-    # -----------------
-    def on_close(self, view):
```

Comment:
The whole on_close/on_pre_close stuff was not necessary, removed. To prevent trouble when closing CodeMap group, I move it to the first group and I close it there, then I remove the group safely, because otherwise there can be interferences if the last file in a pane is closed. I had my layout collapsed, but it didn't happen in a portable installation or before ST3. This way is safer anyway, and there's less useless stuff.